cosmwasm/ibc-union/app/ucs06-funded-dispatch/src/contract.rs

```diff
@@ -0,0 +1,219 @@
+    state::{CONFIG, EXECUTING_PARAMS},
+    types::{FundedDispatchParameters, StoredFundedDispatchFund, StoredFundedDispatchParameters},
+    ContractError,
+};
+
+pub const EXECUTE_REPLY_ID: u64 = 0x1337;
+
+/// Instantiate `ucs06-funded-dispatch`.
+pub fn init(deps: DepsMut, _: Env, msg: InitMsg) -> Result<Response, ContractError> {
+    CONFIG.save(deps.storage, &msg.config)?;
+    Ok(Response::new())
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct MigrateMsg {}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(_: DepsMut, _: Env, _: MessageInfo, _: ()) -> StdResult<Response> {
+    panic!("this contract cannot be instantiated directly, but must be migrated from an existing instantiated contract.");
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(
+    deps: DepsMut,
+    env: Env,
+    msg: UpgradeMsg<InitMsg, MigrateMsg>,
+) -> Result<Response, ContractError> {
+    msg.run(
+        deps,
+        |deps, init_msg| {
+            let res = init(deps, env, init_msg)?;
+            Ok((res, None))
+        },
+        |_deps, _migrate_msg, _current_version| Ok((Response::new(), None)),
+    )
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _: Env,
+    _: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::OnZkgm { message, .. } => internal_execute(deps, message, false),
+        ExecuteMsg::OnIntentZkgm { message, .. } => internal_execute(deps, message, true),
+        ExecuteMsg::Execute { message, intent } => internal_execute(deps, message, intent),
+    }
+}
+
+fn internal_execute(
+    deps: DepsMut,
+    message: Bytes,
+    intent: bool,
+) -> Result<Response, ContractError> {
+    let params = FundedDispatchParameters::abi_decode_params(&message, true)?;
+    if EXECUTING_PARAMS.exists(deps.storage) {
+        return Err(ContractError::ForbiddenReentrancy);
+    }
+    if intent && !params.allow_market_maker() {
+        return Err(ContractError::MarketMakerNotAllowed);
+    }
+    let contract_address =
+        deps.api
+            .addr_validate(std::str::from_utf8(&params.contractAddress).map_err(|_| {
+                ContractError::InvalidContractAddress {
+                    address: params.contractAddress.to_vec().into(),
+                }
+            })?)?;
+    let mut messages = Vec::<CosmosMsg>::with_capacity(1 + params.funds.len());
+    let mut funds = Vec::new();
+    let mut stored_funds = Vec::new();
+    for fund in params.funds.iter() {
+        let denom = std::str::from_utf8(&fund.token).map_err(|_| ContractError::InvalidDenom {
+            denom: fund.token.to_vec().into(),
+        })?;
+        let amount = u128::try_from(fund.amount)
+            .map_err(|_| ContractError::InvalidAmount {
+                amount: fund.amount,
+            })?
+            .into();
+        if deps.querier.query_wasm_contract_info(denom).is_ok() {
```

Comment:
would be better if we query cw20 token info to make sure that the denom is cw20. It's still fine to do it this way though since `IncreaseAllowance` will fail anyways otherwise.

Reply:
I think it's best to use the increase allowance because it's then the only interface required for a contract to be able to be used. It means the contract could be custom (not cw20) and still be used in the process.